js/ui/grid_core/ui.grid_core.editing.js

```diff
@@ -1655,130 +1680,151 @@ const EditingController = modules.ViewController.inherit((function() {
+                });
+            }
+        },
 
-                        dataSource && dataSource.endLoading();
+        _endSaving: function(changes, editData, deferred) {
+            const editMode = getEditMode(this);
+            const dataSource = this._dataController.dataSource();
 
-                        const refreshMode = this.option('editing.refreshMode');
-                        const isFullRefresh = refreshMode !== 'reshape' && refreshMode !== 'repaint';
+            if(editMode !== EDIT_MODE_CELL) {
+                this._resetModifiedClassCells(editData);
+                this._resetEditIndices();
+            }
 
-                        if(!isFullRefresh) {
-                            dataController.push(changes);
-                        }
+            if(editMode === EDIT_MODE_POPUP && this._editPopup) {
+                this._editPopup.hide();
+            }
 
-                        when(dataController.refresh({
-                            selection: isFullRefresh,
-                            reload: isFullRefresh,
-                            load: refreshMode === 'reshape',
-                            changesOnly: this.option('repaintChangesOnly')
-                        })).always(() => {
-                            this._fireSaveEditDataEvents(editData);
-                        }).done(() => {
-                            afterSaveEditData();
-                        }).fail((error) => {
-                            afterSaveEditData(error);
-                        });
-                    } else {
-                        dataSource && dataSource.endLoading();
-                        result.resolve();
-                    }
-                }).fail(function(error) {
-                    dataSource && dataSource.endLoading();
-                    result.resolve(error);
-                });
+            dataSource?.endLoading();
 
-                return result.always(() => {
-                    this._focusEditingCell();
-                    this._saving = false;
-                }).promise();
-            }
+            this._refreshDataControllerAfterSaveEditData(changes, editData, deferred);
+        },
+
+        _cancelSaving: function() {
+            const editMode = getEditMode(this);
+            const dataController = this._dataController;
 
             if(isRowEditMode(this)) {
                 if(!this.hasChanges()) {
-                    this.cancelEditData();
+                    this._cancelEditDataCore();
                 }
             } else if(this.isCellOrBatchEditMode()) {
-                resetEditIndices();
+                if(editMode !== EDIT_MODE_CELL) {
+                    this._resetEditIndices();
+                }
+
                 dataController.updateItems();
             } else {
                 this._focusEditingCell();
             }
 
-            afterSaveEditData();
-            return result.promise();
+            this.executeAction('onSaved');
```

Comment:
Is onSaved event should be fired if saving is canceled?

Reply:
No.